### PR TITLE
Port check endpoint error handling

### PR DESCRIPTION
**Pull request by ijunaidfarooq**

### CODE REVIEW

File: web/controllers/camera_controller.ex [hunk 1]
```diff
@@ -111,4 +111,37 @@ defmodule EvercamMedia.CameraController do
...
+  end
+  defp check_params(params) do
+    with :ok <- valid?("address", params["address"]),
+         :ok <- valid?("port", params["port"]),
+         do: :ok
+  end
+  defp valid?(key, value) when value in [nil, ""],  do: invalid(key)
+  defp valid?("address" = key, value) do
+    case :inet_parse.address(String.to_char_list(value)) do
+       {:ok, _} -> :ok
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: This is indented 3 spaces, should be 2.

File: web/controllers/camera_controller.ex [hunk 2]
```diff
@@ -111,4 +111,37 @@ defmodule EvercamMedia.CameraController do
         send_resp(conn, 500, "Invalid token.")
     end
   end
+
+  #### All defs for params coming from port-check endpoint
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Remove this comment block.

**Comment by mosic** [comment]
Nice work Junaid, the code looks almost ready for merging. You just need to give the private functions some breathing room, group just the ones with the same name.

Look at https://github.com/evercam/evercam-media/blob/master/lib/evercam_media/snapshot/db_handler.ex for an example of this.

**Comment by ijunaidfarooq** [comment]
Thanks!
I have done this on both PRs. Hope fully its done!



Regards


*Junaid Farooq*
Software Developer, Evercam.io
+92347-8666955 | junaid@evercam.io | Evercam <http://evercam.io/> | Skype:
darulehsan1 <https://mail.google.com/mail/u/0/#>
<http://www.facebook.com/iOkay.uOkay>  <http://pk.linkedin.com/in/iamjunnii>
  <http://github.com/ijunaidfarooq>

**Comment by mosic** [comment]
Good work, I rebased it into master. I forgot to mention on your PRs, it would be much nicer to squash the commits into one (so unimportant ones like fixing typos are committed). Also, when updating your branch to master it's better to rebase on top of master then to merge master into it, it's much cleaner.

Regarding the PR: I made some changes to it, most notably:
* Made the `port_open?` function public, so we can use it for https://github.com/evercam/evercam-media/issues/231
* Fixed the port check to include all ports between (and including) 1 and 65535
* Added support for using domains, not just IP addresses
* Used stricter IP address check (e.g. 84.215.67 was valid before, now it's not)

This could have been avoided if you added a couple of unit tests ;)

**Comment by ijunaidfarooq** [comment]
Thanks.. 
Out of interest.. wasnt it better to use new `with else` here? istead of using `cond`?

https://github.com/evercam/evercam-media/blob/master/web/controllers/camera_controller.ex#L123-L129

```
  defp valid?("address", value) do
    cond do
      :inet_parse.strict_address(String.to_char_list(value)) |> elem(0) == :ok -> :ok
      :inet_parse.domain(String.to_char_list(value)) -> :ok
      true -> invalid("address")
    end
  end
```

**Comment by ijunaidfarooq** [comment]
```
defp valid?("addresss", value) do
  with
       :ok <- String.to_char_list(value) |> :inet_parse.strict_address |> elem(0)
       :ok <- String.to_char_list(value) |> :inet_parse.domain do
       :ok
  else
      invalid("address")
  end
end
```

something like that? (not tested)

**Comment by marcoherbst** [comment]
Great stuff. Nice one @mosic @ijunaidfarooq

**Comment by ijunaidfarooq** [comment]
may this 
```
 defp valid?("address", value) do
    cond do
      :inet_parse.strict_address(String.to_char_list(value)) |> elem(0) == :ok -> :ok
      :inet_parse.domain(String.to_char_list(value)) -> :ok
      true -> invalid("address")
    end
  end
```
can be written as 
```
defp valid?("address", value) do
  correct_address = :inet_parse.strict_address(String.to_char_list(value)) |> elem(0) == :ok
  correct_domain  = :inet_parse.domain(String.to_char_list(value))

  if correct_address && correct_domain, do: :ok, else: invalid("address")
end
```
to avoid the `cond` which one you prefer as best to use? @mosic

**Comment by mosic** [comment]
@ijunaidfarooq That wouldn't work as the address couldn't be both a valid IP address and a valid domain.

Btw I missed it in the original review, but `validate` would be a better name than `valid?`, as the convention in Elixir is that functions with `?` should return a boolean value. I renamed it in https://github.com/evercam/evercam-media/commit/d54b84f5eb18e4883010a241b07a6829e312feea.

**Comment by mosic** [comment]
@ijunaidfarooq More information about the naming conventions: http://elixir-lang.org/docs/master/elixir/naming-conventions.html

**Pull request closed without merging**